### PR TITLE
plugin/forward using pkg/up

### PR DESCRIPTION
**Pull request by miekg**
This plumbs through pkg/up to run only 1 HC at any one time; it did simplify code, so that's good.
Couldn't resist other cleanups like remove the `host`: Forward -> Proxy -> host`, to just `Forward -> Proxy`.

### CODE REVIEW

**Comment by miekg** [comment]
Running this on my k8s cluster with a *forward* to 9.9.9.9 using tls

**Comment by miekg** [comment]
And of course my sh*t router does this on healthchecks:
~~~
yment-7d575fbd9-kbmzn[coredns]: 2018/02/07 18:39:51 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.1.67:45004->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-kbmzn[coredns]: 2018/02/07 18:39:51 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.1.67:50820->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-kbmzn[coredns]: 2018/02/07 18:39:51 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.1.67:41097->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-r9rzd[coredns]: 2018/02/07 18:39:52 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.2.72:52371->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-r9rzd[coredns]: 2018/02/07 18:39:52 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.2.72:34127->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-bqdsl[coredns]: 2018/02/07 18:39:52 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.3.56:53378->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-bqdsl[coredns]: 2018/02/07 18:39:52 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.3.56:53395->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-bqdsl[coredns]: 2018/02/07 18:39:52 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.3.56:52071->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-kbmzn[coredns]: 2018/02/07 18:39:53 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.1.67:46823->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-bqdsl[coredns]: 2018/02/07 18:39:54 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.3.56:58615->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-bqdsl[coredns]: 2018/02/07 18:39:54 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.3.56:58120->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-bqdsl[coredns]: 2018/02/07 18:39:54 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.3.56:48363->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-s88x8[coredns]: 2018/02/07 18:39:54 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.1.66:34476->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-kbmzn[coredns]: 2018/02/07 18:39:55 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.1.67:57333->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-kbmzn[coredns]: 2018/02/07 18:39:55 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.1.67:59497->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-kbmzn[coredns]: 2018/02/07 18:39:55 [INFO] healtheck of 192.168.1.1:53 failed with read udp 10.244.1.67:34716->192.168.1.1:53: i/o timeout
default/coredns-deployment-7d575fbd9-kbmzn[coredns]: 2018/02/07 18:39:55 [INFO] healtheck 
~~~
Should be fixed with `max_fails 0` double checking that - should actually also a test.

**Comment by miekg** [comment]
HC metrics look weird: https://snapshot.raintank.io/dashboard/snapshot/XRioEwA3Oe1uK7zsYqguXbfFRIoQwn9T?orgId=2

**Comment by miekg** [comment]
Still don't fully understand what I'm seeing here.

**Comment by miekg** [comment]
This can use a review from someone, /cc @johnbelamaric @chrisohaver (and/or someone else)

File: plugin/forward/README.md [hunk 1]
```diff
@@ -6,10 +6,17 @@
...
 ## Description
 
-The *forward* plugin is generally faster (~30+%) than *proxy* as it re-uses already opened sockets
-to the upstreams. It supports UDP, TCP and DNS-over-TLS and uses inband health checking that is
-enabled by default.
-When *all* upstreams are down it assumes healtchecking as a mechanism has failed and will try to
+The *forward* plugin re-uses already opened sockets to the upstreams. It supports UDP, TCP and
+DNS-over-TLS and uses inband health checking.
+
+When it detects an error an health check is performed. This checks runs in a loop, every *0.5s*, for
```
Comment (chrisohaver) [inline plugin/forward/README.md]: an -> a ... "... a health check ..."

File: plugin/forward/README.md [hunk 2]
```diff
@@ -6,10 +6,17 @@
...
+a long as the upstream reports unhealthy. Once healthy we stop health checking (until the next
```
Comment (chrisohaver) [inline plugin/forward/README.md]: a -> as ... " ... as long as ... "

File: plugin/forward/README.md [hunk 3]
```diff
@@ -6,10 +6,17 @@
...
+error). The health checks use a recursive DNS query (`. IN NS`) to get upstream health. Any response
+that is not an network error (REFUSED, NOTIMPL, SERVFAIL, etc) is taken as a healthy upstream. The
```
Comment (chrisohaver) [inline plugin/forward/README.md]: an -> a  "... a network error ..."

File: plugin/forward/README.md [hunk 4]
```diff
@@ -6,10 +6,17 @@
...
+health check uses the same protocol as specific in **TO**. If `max_fails` is set to 0, no checking
```
Comment (chrisohaver) [inline plugin/forward/README.md]: specific -> specified

File: plugin/forward/README.md [hunk 5]
```diff
@@ -6,10 +6,17 @@
...
+is performed and upstreams will ever be considered unhealthy.
```
Comment (chrisohaver) [inline plugin/forward/README.md]: ever -> never ... "... never be considered unhealthy." ... ?

or perhaps "... always be considered healthy."  (no double negative)

File: plugin/forward/README.md [hunk 6]
```diff
@@ -125,16 +122,10 @@ Proxy everything except `example.org` using the host's `resolv.conf`'s nameserve
...
-
-~~~ corefile
-. {
-    forward . [::1]:1053
-}
-~~~
-
 Proxy all requests to 9.9.9.9 using the DNS-over-TLS protocol, and cache every answer for up to 30
-seconds.
+seconds. Note the `tls_servername` is mandatory if you want to working setup, as 9.9.9.9 can't be
```
Comment (chrisohaver) [inline plugin/forward/README.md]: to -> a  " ... a working setup ..."

File: plugin/forward/forward.go [hunk 7]
```diff
@@ -92,7 +92,27 @@ func (f *Forward) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg
...
+				ret.SetReply(r)
+				ret.Truncated = true
+				ret.Authoritative = true
+				ret.Rcode = dns.RcodeSuccess
+			}
+			err = nil // and reset err to pass this back to the client.
+		}
+
 		if err != nil {
+			// Kick of health check to see if *our* upstream is broken.
```
Comment (chrisohaver) [inline plugin/forward/forward.go]: "Kick of" -> "Kick off"

**Review by chrisohaver — COMMENTED** [review]
LGTM.  I suggested some minor grammatical fixes to README.

**Comment by chrisohaver** [comment]
> HC metrics look weird

You mean that there is "noise" (not flat steps)?  Variability in sample collection?

**Comment by miekg** [comment]
metrics still is weird (at home) - which I should double check (and I will). But the hc-on-failure is a **so much** better model than the HC-every-0.5s that this should go in for 1.0.6

Thanks for the review!

**Review by chrisohaver — APPROVED** [review]

**Pull request merged**